File: python/55-Jump Game.py
from typing import List
from functools import lru_cache

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        "Top-Down with Memoization, Time: O(n^2), Space: O(n), TLE"
        n = len(nums)
        @lru_cache(None)
        def helper(i):
            "True if we can arrive last when we start at i"
            if i == n-1: return True
            if i < 0 or i >=n: return False
            for j in range(1, nums[i]+1):
                if helper(i+j): return True
            return False
        return helper(0)
    def canJump(self, nums: List[int]) -> bool:
        "Down-Top DP, Time: O(n^2), Space: O(n), TLE"
        n = len(nums)
        dp = ["unknown"]*n
        dp[-1] = "True"
        for i in reversed(range(n-1)):
            for j in range(1, nums[i]+1):
                if dp[i+j] == "True":
                    dp[i] = "True"
                    break
            if dp[i] != "True":
                dp[i] = "False"
        return True if dp[0] == "True" else False
    def canJump(self, nums: List[int]) -> bool:
        "Improvement of above, Time: O(n), Space: O(1)"
        "Observation from above: see the break statement, that tell us that we only need the left most True position"
        n = len(nums)
        left_T_pos = n-1
        for i in reversed(range(n-1)):
            if i + nums[i] >= left_T_pos:
                left_T_pos = i
        return left_T_pos == 0

nums = [2,3,1,1,4]
result = Solution().canJump(nums)
assert result == True

nums = [3,2,1,0,4]
result = Solution().canJump(nums)
assert result == False

File: cpp/55. Jump Game.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

// class Solution {
// /*
// dp[i] = any(dp[j] for j in i+1...i+nums[i])
// dp[n-1] = true
// ans = dp[0]
// */
// public:
//     bool canJump(vector<int>& nums) {
//         // Time: O(n^2), Space: O(n)
//         int n = nums.size();
//         int dp[n]; fill(dp, dp+n, false); dp[n-1] = true;
//         for(int i = n-2; i >= 0; i--) {
//             for(int j = i+nums[i]; j > i; j--) {
//                 if(j<n && dp[j]) {
//                     dp[i] = true;
//                     break;
//                 }
//             }
//         }
//         return dp[0];
//     }
// };

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int maxReachable = 0;
        int n = nums.size();
        for(int i = 0; i < n; ++i) {
            if(i>maxReachable) return false;
            maxReachable = max(maxReachable, i+nums[i]);
        }
        return true;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi nums; bool ans;

    solution = Solution();
    nums = {2,3,1,1,4};
    ans = solution.canJump(nums);
    assert(ans==true);

    solution = Solution();
    nums = {3,2,1,0,4};
    ans = solution.canJump(nums);
    assert(ans==false);

    solution = Solution();
    nums = {2, 0};
    ans = solution.canJump(nums);
    assert(ans==true);
}